File: events/2015/19/solution.rs
use aoc_utils::read_input_as_array;
use aoc_utils::measure_performance;
use aoc_utils::AOCDayResults;
use aoc_utils::AOCTimestamp;
use chrono::{DateTime, Utc};

use std::collections::{HashMap, HashSet};

fn find_substitutes(medicine: String, molecule: String, replacement: String) -> Vec<String> {
  let med_len = medicine.chars().count();
  let mol_len = molecule.chars().count();

  if med_len < mol_len {
    return Vec::new();
  }

  let mut new_medicine: String = "".to_string();
  let mut found_medicines: Vec<String> = Vec::new();

  for i in 0..med_len-mol_len+1 {
    let test_part = &medicine[i..i+mol_len];

    if test_part == molecule {
      let new = new_medicine.clone() + &replacement + &medicine[i+mol_len..];
      found_medicines.push(new);
    }
    new_medicine += &medicine[i..i+1];
  }
  found_medicines
}
// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
fn part1(input: &Vec<String>) -> i32 {
  let mut replacements: HashMap<String, Vec<String>> = HashMap::new();
  let mut medicine: String = "".to_string();
  let mut found_medicines: HashSet<String> = HashSet::new();

  for (_i, line) in input.iter().enumerate() {
    if line.is_empty() {
      continue;
    }

    if line.contains("=>") {
      let parts: Vec<&str> = line.split(" => ").collect::<Vec<&str>>();

      replacements.entry(parts[0].to_string()).or_insert_with(Vec::new).push(parts[1].to_string());
    } else {
      medicine = line.to_string();
    }
  }

  for (molecule, value) in &replacements {
    for substitute in value {
      let results = find_substitutes(medicine.clone(), molecule.to_string(), substitute.to_string());

      for res in results {
        found_medicines.insert(res);
      }
    }
  }

  found_medicines.len().try_into().unwrap()
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
fn part2(input: &Vec<String>) -> i32 {
  let mut replacements: Vec<(String, String)> = Vec::new();
  let mut medicine: String = "".to_string();

  for line in input {
    if line.is_empty() {
      continue
    }

    if line.contains("=>") {
      let parts: Vec<&str> = line.split(" => ").collect::<Vec<&str>>();

      let from = parts[1].chars().rev().collect();
      let to = parts[0].chars().rev().collect();

      replacements.push((from, to));
    } else {
      medicine = line.chars().rev().collect();
    }
  }

  let mut steps = 0;

  while medicine.len() > 1 {
    let mut best_index = usize::MAX;
    let mut best_replacement: Option<&(String, String)> = None;

    for replacement in &replacements {
      let (from, _to) = replacement;
      if let Some(index) = medicine.find(from) {
        if index < best_index {
          best_index = index;
          best_replacement = Some(replacement);
        }
      }
    }

    if let Some((from, to)) = best_replacement {
      medicine = format!("{}{}{}", &medicine[..best_index], to, &medicine[best_index + from.len()..]);
      steps += 1;
    } else {
      panic!("Stuck after {} steps", steps);
    }
  }
  
  Some(steps).unwrap()
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
fn main() {
  // Read the arguments
  let args = std::env::args().collect::<Vec<String>>();
  let default_file_name = String::from("input.txt");
  let file_name = args.get(1).unwrap_or(&default_file_name);

  // Start the timer
  let start_timestamp = std::time::SystemTime::now();
  let start_time = std::time::Instant::now();
  
  // Read the input
  let input = read_input_as_array(file_name).unwrap();

  // Run the Parts
  let p1_result = measure_performance(part1, &input);
  let p2_result = measure_performance(part2, &input);

  // Stop the timer
  let end_time = std::time::Instant::now();
  let end_timestamp = std::time::SystemTime::now();
  let duration = end_time.duration_since(start_time).as_nanos();

  let results = AOCDayResults {
    year: 2015,
    day: 19,
    part1: p1_result,
    part2: p2_result,
    duration: duration,
    timestamp: AOCTimestamp {
      start: DateTime::<Utc>::from(start_timestamp).to_rfc3339(),
      end: DateTime::<Utc>::from(end_timestamp).to_rfc3339(),
    },
  };

  let json_results = serde_json::to_string(&results).unwrap();
  println!("{}", json_results);
}

File: events/2015/19/solution.ts
import { AOCDayResults } from "@repo/utils-ts/returnType";
import { readInputAsArray } from "@repo/utils-ts/readInput";
import { measurePerformance } from "@repo/utils-ts/measurePerformance";

function findSubstitutes(medicine: string, molecule: string, replacement: string): string[] {
  let medLen = medicine.length
  let molLen = molecule.length

  if (medicine.length < molecule.length) {
    return []
  }

  let newMedicine = ""
  let foundMedicines: string[] = []


  for (var i = 0; i < medLen; i++) {
    let testPart = medicine.substring(i, i+molLen)
    
    if (testPart == molecule) {
      foundMedicines.push(newMedicine + replacement + medicine.substring(i+molLen))
    }
    newMedicine += medicine[i]
  }
  
  return foundMedicines
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
function part1(input: string[]): number | undefined {
  let replacements: Record<string, string[]> = {}
  let medicine: string = ""
  let foundMedicines: string[] = [];

  // Loop over the input
  for (var i = 0; i < input.length; i ++) {
    let line = input[i];

    // If the line contains no content, ignore it
    if (line.length == 0) {
      continue;
    }

    // Does the line contain "=>"?
    if (line.includes("=>")) {
      // This line is a replacement definition
      let parts = line.split(" ")

      if (replacements[parts[0]] == undefined) {
        replacements[parts[0]] = [parts[2]]
      } else {
        replacements[parts[0]].push(parts[2])
      }
    } else {
      // This is the medicine line
      medicine = line
    }
  }

  // Loop over the replacements
  for (const molecule in replacements) {
    for (const index in replacements[molecule]) {
      let substitute = replacements[molecule][index]
      
      let results = findSubstitutes(medicine, molecule, substitute);
      
      for (var i = 0; i < results.length; i++) {
        if (!foundMedicines.includes(results[i])) {
          foundMedicines.push(results[i])
        }
      }
    }
  }

  return foundMedicines.length;
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
function part2(input: string[]): number | undefined {
  let replacements: [string, string][] = []
  let medicine: string = ""

  // Parse the input
  for (var i = 0; i < input.length; i++) {
    let line = input[i];

    if (line.length == 0) { continue }

    if (line.includes("=>")) {
      const parts = line.split(" ")

      let newReplacement: [string, string] = [parts[2].split('').reverse().join(''), parts[0].split('').reverse().join('')]
      replacements.push(newReplacement)
    } else {
      medicine = line.split('').reverse().join('')
    }
  }
  
  let steps = 0;

  // Keep reducing until we get to a single element
  while (medicine.length > 1) {
    // Find the leftmost occurrence among all possible replacements
    let bestIndex = Infinity;
    let bestReplacement: [string, string] | null = null;

    for (const replacement of replacements) {
      const [from, to] = replacement;
      const index = medicine.indexOf(from);
      if (index !== -1 && index < bestIndex) {
        bestIndex = index;
        bestReplacement = replacement;
      }
    }

    // If we found a replacement, apply it
    if (bestReplacement !== null) {
      const [from, to] = bestReplacement;
      medicine = 
        medicine.substring(0, bestIndex) +
        to +
        medicine.substring(bestIndex + from.length);
      steps++;
    } else {
      throw new Error(`Stuck after ${steps} steps`);
    }
  }

  return steps;
}


// ----------------------------------------------------------------------------------------------------
// | Solve the puzzle
// ----------------------------------------------------------------------------------------------------
function solve() {
  // Get the arguments
  const args = process.argv.slice(2);
  const fileName = args[0] || "input.txt";

  // Start the timer
  const startTimeStamp = new Date(Date.now()).toISOString();
  const startTime = performance.now();

  // Read the input as an array of strings
  const input = readInputAsArray(fileName);

// Run the parts
const p1Result = measurePerformance(() => part1(input));
const p2Result = measurePerformance(() => part2(input));

// End the timer
const endTimeStamp = new Date(Date.now()).toISOString();
const endTime = performance.now();
const duration = endTime - startTime;

// Return the results
return {
  Year: 2015,
  Day: 19,
  Part1: p1Result,
  Part2: p2Result,
  Duration: Math.round(duration * 1000000), // Convert to nanoseconds
  Timestamp: {
    Start: new Date(startTimeStamp).toISOString(),
    End: new Date(endTimeStamp).toISOString(),
  },
} as AOCDayResults;
  
}
const results = solve();
console.log(JSON.stringify(results, null, 2));